### ISSUE TITLE
When does/did the French Republican year start ?

### ISSUE THREAD

**Issue by catull** [ledahulevogyre/threeten-extra#5]
The French constitution declared three conflicting articles on the beginning of the year, which can be read [here](http://www.gauchemip.org/spip.php?article1090) or [here](http://mjp.univ-perp.fr/france/1793ererep.htm)

> Article premier: L'ère des Français compte de la fondation de la république, qui a eu lieu le 22 septembre 1792 de l'ère vulgaire, jour où le soleil est arrivé à l'équinoxe vrai d'automne, en entrant dans le signe de la balance à 9 heures 18 minutes 30 secondes du matin, pour l'observatoire de Paris.

In short: the year starts with the day of the autumnal equinox as seen from the observatory in Paris.

> Article III: Le commencement de chaque année est fixé à minuit, commençant le jour où tombe l'équinoxe vrai d'automne pour l'observatoire de Paris.

In short: same as article 1, the beginning of  the year is on the day of the autumnal equinox as seen in Paris.

The last article - 16 - though "fixes" the leap-year to every 4th year:

> Article XVI: Tous les quatre ans, ou toutes les Franciades, au jour de la Révolution, il sera célébré des jeux républicains en mémoire de la révolution française.

The equinoxes do not follow this strict rule.

So, there are 4 variants to consider:

1. **equinox**: only obeying article 1 and 3, follow the astronomical event.
    Most years will have 5, some will have 6 complimentary days.
2. **Romme**: 6th complimentary day - _leap day_ - is at the end of every 4th year, in years 4 * n
3. **continuous**: _leap day_ before every 4th year or century starts, thus years 4 * n or 400 * n, but not n * 100 or n * 4000
   -> 3, 7, 11, 15, .., 96, 104, ..., 296, 304, ..., 396, 400, 404, ..., 796, 800, 804, ..., 3996, 4004, ...
4. **128-year**: _leap day_ in years divisible by 4 but not divisible by 128

According to Wikipedia article [French Republican Calendar](https://en.wikipedia.org/wiki/French_Republican_Calendar), variant 2 (Romme) would have had the entire 3rd century (from Gregorian year 1992 - 2091) start each year on September 22nd in the Gregorian Calendar.

Also, variant 4 has the benefit that for the Gregorian years 1920 - 2047, the French Republican years all start on September 23rd.

**Comment by catull** [comment]
We could have 4 different classes, perhaps even 1 main class and 3 sub-classes implementing the different variants.
At any rate, the first variant (**equinox**) seems to be the one legally used.
Thus it must be the main featured one.
The other variants are all simpler to calculate.

**Comment by ledahulevogyre** [comment]
Do you think it's ok to write ofEpochDay() as only dependent of isLeapYear() ? So that subclasses only have to override iLeapYear() ? 
I think it means iterating on all years since year I. So it's not very efficient.

What do you mean by main featured one ? base class ?
Maybe it's more appropriate to use the simplest variant as the base class. And add complexity to the subclasses.

**Comment by catull** [comment]
isLeapYear() will have 4 implementations, that is true.

To ofEpochDay is not going to change, there is a way to keep it generic by implementing 
```java
FrenchRepublicChronology#getLeapYearsBefore(long)
```
Look at [Symmetry454Chronology#getLeapYearsBefore(long)](https://github.com/ThreeTen/threeten-extra/blob/master/src/main/java/org/threeten/extra/chrono/Symmetry454Chronology.java#L506);

That method also works for dates before the Republican Era.

I have an idea how to implement it.

I see two approaches.

**A**) `FrenchRepublicDate` is an abstract class having 4 concrete final classes extend from it:
`FrenchRepublicEquinoxDate`
`FrenchRepublicRommeDate`
`FrenchRepublicContinousDate`
`FrenchRepublicY128Date`

**B**) `FrenchRepublicDate` if final and has a `mode`, which is passed to `FrenchRepublicChronology#date(int, int, int, Mode)`.
Consequently, an instance of  `FrenchRepublicDate` is constructed in 4 different ways depending on the `Mode`.

In either approaches, the equinox variant/mode is the default.
The other variants/modes are simpler to implement.

**Comment by ledahulevogyre** [comment]
getLeapYearsBefore() is a good idea. The subclasses will still have to override two methods instead of one. But it's probably for the better.

I think approach A is better.

**Comment by catull** [comment]
In the end, we might just settle with `equinox` and a second one.

But for that, we need to know which one was everyone's favourite at the time.
Do you know ?
I read that Romme described his variant, but then he was arrested and executed.
So his method was not further used or considered.

In the 14 years the French Republican calendar was used, the equinox method determined when "1 Vendémiaire an N" would be, by observation and consensus.
The moment in time was considered very hard to calculate in advance.

These days, we have astronomical formulae which help us.

I do not have enough evidence which one of the other 3 simpler variants stood out.

From a mathematical point, I would rule out Romme.
It is as imprecise as the Julian calendar.
It ruins the innovation with lower reliability than the Gregorian calendar, which it wants to replace.

Between `continuous` and `128-year`, I would have to calculate the average error they both produce.

In short, good candidates are `equinox` and one of `continuous` or `128-year`.

**Comment by catull** [comment]
In the book "Calendrical Calculations", the authors picked
- `equinox` - calling it `French Revolutionary` calendar
- `continuous` which they named `French Arithmetic` calendar

Don't have a copy of the book with me, but I have to re-read that chapter.
It does not necessarily mean that we come to the same conclusion.

**Comment by ledahulevogyre** [comment]
Well I'm not an historian, but I think Romme's version is definitely the good candidate, for them.

**Comment by catull** [comment]
Perhaps you meant to write:

> Well I'm not an historian, but I think Romme's version is definitely **NOT** the good candidate, for them.

`Them` being the authors of "Calendrical Calculations".

**Comment by catull** [comment]
The problem with either `continuous` and `128-year` variant is that they experience a gap of 8 years between one leap year and the next every now and then, such as around some centuries (continuous) or from year 124 to 132 (128).

With `equinox` the leap years are smoothly spread out, no earlier than 4 AND no later than 5 years,  no exceptions.

**Comment by catull** [comment]
Enough talk, let the code speak for itself
It will take me some time to implement the variants.

**Comment by ledahulevogyre** [comment]
No, I meant Romme's version most certainly interests historians.

The authors of "Calendrical Calculations" are probably more interested in the computation aspect.

**Comment by catull** [comment]
I see, your statement now makes sense.
And yes, the authors are interested in the mathematical aspect, how to implement it.
